### PR TITLE
WIP More precise coverage

### PR DESCRIPTION
**Pull request by carnaval**
Make coverage depends on lowered code statement number instead of text
line numbers. The reported number is the minimum count of every statements
mapping to the relevant line.

```
        - function f(b)
        1     b && (x = 4)
        1     x = 2
        - end

        - f(true)
```

```
        - function f(b)
        0     b && (x = 4)
        1     x = 2
        - end

        - f(false)
```

just hacked this, not tested at all, consider this my argument against #11792 :-)

### CODE REVIEW

**Comment by timholy** [comment]
:+1:

**Comment by ScottPJones** [comment]
Hey, you won't get any argument from me!  I did #11792 just because people said the by line coverage couldn't be fixed...
This is exactly what I wanted originally!
:+1: :100:

File: src/codegen.cpp [hunk 1]
```diff
@@ -1134,8 +1134,10 @@ const jl_value_t *jl_dump_function_asm(void *f)
 
 typedef std::map<std::string,std::vector<GlobalVariable*> > logdata_t;
 static logdata_t coverageData;
+// map file => pc => line no
```
Comment (tkelman) [inline src/codegen.cpp]: what's `pc` ?

Comment (ScottPJones) [inline src/codegen.cpp]: pc = program counter = ip = instruction pointer

**Comment by StefanKarpinski** [comment]
This seems great. Does reporting fractional coverage make sense? I.e. 0.5/1 or something like that.

**Comment by JeffBezanson** [comment]
Love it!

File: src/codegen.cpp (hunk 1 above)
Comment (carnaval) [inline src/codegen.cpp]: yeah the comment is wrong now also :-) and the code a bit ugly, I'll do a cleanup pass tomorrow

**Comment by kshyatt** [comment]
:100: !

**Comment by mschauer** [comment]
How about reporting the range of counts so "partial bottlenecks" are visible (bottlenecks constraint to parts of the line).

**Comment by ScottPJones** [comment]
That sounds like a good idea @mschauer

**Comment by tkelman** [comment]
Right now the .cov format is sort of designed for reporting a single number, right? So either we would have to extend that format to report additional information, or output multiple copies of the file for different choices of a single number to report?

**Comment by ScottPJones** [comment]
I'd figured a simple extension would be done, like `min:max` instead of `count`.
BTW, what is done to handle multiple processes trying to write coverage information at once?

**Comment by mbauman** [comment]
I think the very clever idea here is to keep track (in memory) of the hit counts of every statement.  Then when you go back to output the information for a single line, you take the minimum number from all the statements on that line.  The two code blocks in the first comment are two independent runs from two separate files.

**Comment by IainNZ** [comment]
So we now have codecov.io support, which can track partial coverage. Just to kinda incentivize getting this over the line.

**Comment by waldyrious** [comment]
> So we now have codecov.io support

Noob comment here, but doesn't that mean something should be visible on [this page](https://codecov.io/github/JuliaLang/julia)? Is it just a matter of time, or am I looking at the wrong place?

**Comment by IainNZ** [comment]
Did anyone say it should be? For now its living at https://codecov.io/github/kshyatt/julia?ref=master

**Comment by kshyatt** [comment]
So the buildbots aren't running right now, @waldyrious, for a variety of reasons (ask @tkelman or @staticfloat for more). I've been running coverage on my fork of Julia for now. If anyone would like some help getting a copy of Julia coverage running on their own machines, drop me a line any time.

**Comment by waldyrious** [comment]
Ah, got it, thanks for the clarifications.

**Comment by jiahao** [comment]
bump

**Comment by Keno** [comment]
This is somewhat stale, but @vtjnash has done some work on more precise coverage recently, which I think had roughly the same effect.

**Comment by vtjnash** [comment]
This is a significantly different piece of design from my recent work. Though I agree we probably can’t use the PR in its current state.

**Pull request closed without merging**